### ISSUE TITLE
What's the most efficient way to bind many arrays?

### ISSUE THREAD

**Issue by vchrn** [oracle/odpi#78]
Good afternoon,

It's just a question, not an issue. An application we've inherited writes records into a very large historical object table, over time the object definition evolved into a beast with more than 800 columns, many of them are of `VARRAY` type. Currently the application uses something like `INSERT INTO table VALUES(:record)` to insert new rows, where `:record` is an object.

Unfortunately, calling `dpiObject_setAttributeValue` for so many attributes takes approximately 100 milliseconds per record, so the rate of insertions is very low. I thought that maybe `INSERT INTO table (column1, column2, ...) VALUES(:column1, :column2, ...)` with a (hopefully) quicker refresh of `dpiData` values will be faster, but with this syntax I can't find a way to bind `dpiConn_newVar`-created arrays into `VARRAY` columns. Is there an efficient way to do so? Or maybe there is a better way of speeding things up?

Thanks in advance for your help.

**Comment by anthony-tuininga** [comment]
If you have to set all (or most) of the columns you could try the following instead:

```python
args = [arg1, arg2, arg3, arg4, ....arg821]
cursor.execute("insert into table values (my_record(:1, :2, :3, :4, :5, ....., :821))", args)
```

I'm not sure if that will be faster or not but it is an option you could experiment with. If something else comes to mind I'll let you know!

**Comment by vchrn** [comment]
Yeah, plus a PL/SQL block around it, otherwise it will complain about arrays. And another question about binding arrays then :)

```sql
create type my_test_nested as table of number;
create type my_test_varray as varray(2) of number;

create table my_test_table (field my_test_varray);

declare
    procedure my_test_insert(p_var_nested in my_test_nested) is
        p_var_varray my_test_varray;
    begin
        select * bulk collect into p_var_varray from table(p_var_nested);
        insert into my_test_table (field) values (p_var_varray);
    end;
begin
    my_test_insert(my_test_nested(1, 2));
end;
```

Let's assume that I do `my_test_insert(:local_array)` instead. As far as I can see, what `dpiConn_newVar` gives us for a plain array is essentially a nested table, while `field` is a `VARRAY`. Is there a better way to assign `my_test_nested(1, 2)` (or any array that comes from the C world) into `my_test_table.field`?

**Comment by anthony-tuininga** [comment]
Looks like I answered your questions, so I'll close this now. Feel free to re-open if you have another related question.

**Comment by vchrn** [comment]
Sorry for a late reply. I've managed to achieve good performance with binding plain C arrays and then using generated conversion functions to unnest them into VARRAY columns in PL/SQL.

You've asked for a performance test though, and I've composed an [artificial example](https://gist.github.com/vchrn/b72961419c6b730f9bdd2254ebe9f0af). There is clearly something very inefficient happening in `dpiObject_setAttributeValue`.

```sh
# With the code as seen in the gist.
> ./test
100 mapping loops took 11088 ms.

# With dpiObject_setAttributeValue commented out.
> ./test
100 mapping loops took 55 ms.
```

**Comment by vchrn** [comment]
We use `instantclient_18_3` on Linux, everything is compiled at `-O3`.

**Comment by anthony-tuininga** [comment]
Thanks. I just tried your artificial example and I see the same sort of behaviour -- about 100 times longer with the call to dpiObject_setAttributeValue(). I'll see if I can discover why that is and get back to you.

**Comment by felipenoris** [comment]
@vchrn , I would like to know how this solution compares to using `dpiStmt_executeMany` and binding the arrays to variables. Have you tried that?
